core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -598,14 +599,24 @@ private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting fail
                     while (initializingNode.hasNext()) {
                         ShardRouting shardRouting = initializingNode.next();
                         if (shardRouting.isRelocationTargetOf(failedShard)) {
-                            logger.trace("{} is removed due to the failure of the source shard", shardRouting);
-                            initializingNode.remove();
+                            if (failedShard.primary()) {
+                                logger.trace("{} is removed due to the failure of the source shard", shardRouting);
+                                initializingNode.remove();
+                            } else {
+                                logger.trace("{}, relocation source failed, mark as initializing without relocation source", shardRouting);
+                                removedRelocationSourceOnRelocationTarget = true;
+                                initializingNode.removeRelocationSource();
+                            }
+                            break;
                         }
                     }
                 }
             }
-
-            matchedNode.moveToUnassigned(unassignedInfo);
+            if (removedRelocationSourceOnRelocationTarget) {
```

Comment:
The current logic in this method is a bit weird.
When a shard is found that matched `failedShard` we first call `matchedNode.remove()` (earlier up in the method, not shown by Github diff here). This means that the shard gets removed. This makes sense for example if the failed shard is a relocation target. If it is a started shard, however, we need to readd it as unassigned shard, which happens two lines below this comment.

The case we want to cover in this PR is that we remove relocation source (not add it back as unassigned shard) but keep the relocation target.